src/cmd/builtin/dmesg.c

```diff
@@ -25,9 +82,190 @@ static struct optparse_option dmesg_opts[] = {
+{
+    char *extra;
+    struct tm gm_tm;
+    time_t t;
+
+    memset (tm, 0, sizeof (*tm));
+
+    if (!(extra = strptime (s, "%C%y-%m-%dT%H:%M:%S", &gm_tm)))
+        return -1;
+
+    if ((t = portable_timegm (&gm_tm)) == (time_t) -1)
+        return -1;
+
+    if (!(localtime_r (&t, tm)))
+        return -1;
+    tv->tv_sec = t;
+
+    if (extra[0] == '.') {
+        char *endptr;
+        double d = strtod (extra, &endptr);
+
+        /*  In this implementation, the rest of the timestamp must be
+         *   "Z" to indicate UTC or "Zulu time":
+         */
+        if (endptr == NULL || !streq (endptr, "Z"))
+            return -1;
+        tv->tv_usec = d * 1000000;
+    }
+    return 0;
+}
+
+static double tv_to_double (struct timeval *tv)
+{
+    return (tv->tv_sec + (tv->tv_usec/1e6));
+}
+
+static const char *months[] = {
+    "Jan",
+    "Feb",
+    "Mar",
+    "Apr",
+    "May",
+    "Jun",
+    "Jul",
+    "Aug",
+    "Sep",
+    "Oct",
+    "Nov",
+    "Dec",
+    NULL
+};
+
+void print_human_timestamp (struct dmesg_ctx *ctx, struct stdlog_header hdr)
+{
+    struct tm tm;
+    struct timeval tv;
+    if (parse_timestamp (hdr.timestamp, &tm, &tv) < 0) {
+        printf ("%s[%s]%s ",
+                dmesg_color (ctx, DMESG_COLOR_TIME),
+                hdr.timestamp,
+                dmesg_color_reset (ctx));
+    }
+    if (tm.tm_year == ctx->last_tm.tm_year
+        && tm.tm_mon == ctx->last_tm.tm_mon
+        && tm.tm_mday == ctx->last_tm.tm_mday
+        && tm.tm_hour == ctx->last_tm.tm_hour
+        && tm.tm_min == ctx->last_tm.tm_min) {
+        /*  Within same minute, print offset in sec */
+        double dt = tv_to_double (&tv) - tv_to_double (&ctx->last_tv);
+        printf ("%s[%+11.6f]%s ",
+                dmesg_color (ctx, DMESG_COLOR_TIME),
+                dt,
+                dmesg_color_reset (ctx));
+        if (ctx->delta)
+            ctx->last_tv = tv;
+    }
+    else {
+        /* New minute, print datetime */
+        printf ("%s[%s%02d %02d:%02d]%s ",
+                dmesg_color (ctx, DMESG_COLOR_TIMEBREAK),
+                months [tm.tm_mon],
+                tm.tm_mday,
+                tm.tm_hour,
+                tm.tm_min,
+                dmesg_color_reset (ctx));
+        ctx->last_tv = tv;
+        ctx->last_tm = tm;
+    }
+}
+
+static const char *severity_color (struct dmesg_ctx *ctx, int severity)
+{
+    switch (severity) {
+        case LOG_EMERG:
+            return dmesg_color (ctx, DMESG_COLOR_EMERG);
+        case LOG_ALERT:
+            return dmesg_color (ctx, DMESG_COLOR_ALERT);
+        case LOG_CRIT:
+            return dmesg_color (ctx, DMESG_COLOR_CRIT);
+        case LOG_ERR:
+            return dmesg_color (ctx, DMESG_COLOR_ERR);
+        case LOG_WARNING:
+            return dmesg_color (ctx, DMESG_COLOR_WARNING);
+        case LOG_DEBUG:
+            return dmesg_color (ctx, DMESG_COLOR_DEBUG);
+        default:
+            return "";
```

Comment:
Were LOG_NOTICE and LOG_INFO left off on purpose?

On my system DEBUG, INFO, and NOTICE all look the same.

Reply:
Yes, LOG_NOTICE and LOG_INFO are left in the default color. I could add them to be more explicit if that is preferrable. On my system LOG_DEBUG is gray compared to white for the default text. (e.g. see above screenshots)

So if you run this:
```
$ printf "hello\n\x1b[37mhello\x1b[0m\n"
```

Both lines look the same? I guess this is why util-linux supports [`terminal-colors.d(5)`](https://manpages.ubuntu.com/manpages/xenial/man5/terminal-colors.d.5.html). I was thinking one day we could support that, but it seemed like overkill for now.

Maybe for now we could try this?

```diff
diff --git a/src/cmd/builtin/dmesg.c b/src/cmd/builtin/dmesg.c
index f349a18a3..c9db85950 100644
--- a/src/cmd/builtin/dmesg.c
+++ b/src/cmd/builtin/dmesg.c
@@ -69,7 +69,7 @@ static const char *dmesg_colors[] = {
     [DMESG_COLOR_CRIT]          = ANSI_COLOR_BOLD ANSI_COLOR_RED,
     [DMESG_COLOR_ERR]           = ANSI_COLOR_RED,
     [DMESG_COLOR_WARNING]       = ANSI_COLOR_BOLD,
-    [DMESG_COLOR_DEBUG]         = ANSI_COLOR_GRAY,
+    [DMESG_COLOR_DEBUG]         = ANSI_COLOR_HALFBRIGHT ANSI_COLOR_BLUE,
 };
```

Reply:
Yeah pretty close to the same.  One's a tiny bit brighter.

![Screenshot from 2022-04-19 09-56-22](https://user-images.githubusercontent.com/169947/164056191-ac392d7b-faa5-4233-8e2c-674b05bbb9bd.png)

With your changes - better!

![Screenshot from 2022-04-19 09-58-56](https://user-images.githubusercontent.com/169947/164056496-90cd9ab6-7293-447d-8bbe-f1db443795ea.png)

Reply:
Uh, halfbright blue was unreadable in iTerm on my Mac, so that probably isn't great.

Reply:
I'll push a change with just normal blue and see how that goes. Looks fine on the two terminals I tried. Not sure if it ends up being too busy though.

Reply:
Ok, I pushed the result. Maybe try this

```sh
for severity in emerg alert crit err warning notice info debug;
    do flux logger --severity=$severity This is a message at level $severity;
done
```
And ensure it looks ok on your terminal with `flux dmesg -H`

Reply:
That looks good!

![Screenshot from 2022-04-19 10-15-36](https://user-images.githubusercontent.com/169947/164059277-c01b9a94-fbdf-4268-abb5-fcda56dec608.png)

Reply:
Ok, I squashed the color change and force pushed. Hope that was ok!